sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -0,0 +1,230 @@
+        /// <param name="partitionProcessorFactory">A factory used to create partition processors.  Its implementation must be provided by the user.</param>
+        /// <param name="partitionManager">A partition manager used to interact with the storage system.  Its implementation must be provided by the user.</param>
+        ///
+        public EventProcessor(string consumerGroup,
+                              EventHubClient eventHubClient,
+                              IPartitionProcessorFactory partitionProcessorFactory,
+                              IPartitionManager partitionManager) : this(consumerGroup, eventHubClient, partitionProcessorFactory, partitionManager, null)
+        {
+        }
+
+        /// <summary>
+        ///   Initializes a new instance of the <see cref="EventProcessor"/> class.
+        /// </summary>
+        ///
+        /// <param name="consumerGroup">The name of the consumer group this event processor is associated with.  Events are read in the context of this group.</param>
+        /// <param name="eventHubClient">The client used to interact with the Azure Event Hubs service.</param>
+        /// <param name="partitionProcessorFactory">A factory used to create partition processors.  Its implementation must be provided by the user.</param>
+        /// <param name="partitionManager">A partition manager used to interact with the storage system.  Its implementation must be provided by the user.</param>
+        /// <param name="options">The set of options to use for this event processor.</param>
+        ///
+        public EventProcessor(string consumerGroup,
+                              EventHubClient eventHubClient,
+                              IPartitionProcessorFactory partitionProcessorFactory,
+                              IPartitionManager partitionManager,
+                              EventProcessorOptions options)
+        {
+            Guard.ArgumentNotNull(nameof(eventHubClient), eventHubClient);
+            Guard.ArgumentNotNullOrEmpty(nameof(consumerGroup), consumerGroup);
+            Guard.ArgumentNotNull(nameof(partitionProcessorFactory), partitionProcessorFactory);
+            Guard.ArgumentNotNull(nameof(partitionManager), partitionManager);
+
+            InnerClient = eventHubClient;
+            ConsumerGroup = consumerGroup;
+            PartitionProcessorFactory = partitionProcessorFactory;
+            PartitionManager = partitionManager;
+            Options = options?.Clone() ?? new EventProcessorOptions();
+
+            InstanceId = Guid.NewGuid().ToString();
+        }
+
+        /// <summary>
+        ///   Starts the event processor.  In case it's already running, nothing happens.
+        /// </summary>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        public async Task Start()
+        {
+            if (RunningTask == null)
```

Comment:
Why would we need the outer if? To avoid unnecessarily blocking the thread?

Reply:
The majority case is "this is already running" and synchronization is expensive.  The pattern is intended to say "normally, this won't be null, if it is then we have to do some work" .. but it may have changed in the time it took me to get the lock, so lets make sure we need to do the work still. 

[https://en.wikipedia.org/wiki/Double-checked_locking](https://en.wikipedia.org/wiki/Double-checked_locking) goes into more detail.